### PR TITLE
Use tapir for WakeApi

### PR DESCRIPTION
**Pull request by stsatlantis**
Provided tests
Fixed when there is no cache hit caused NPE

### CODE REVIEW

File: src/main/scala/com/wanari/renervator/api/WakerApi.scala [hunk 1]
```diff
@@ -1,40 +1,71 @@
...
-  case class HostDTO(id: Long, name: String, isOnline: Boolean)
+  sealed abstract class ErrorResponse(message: Option[String]) extends Product with Serializable
+
+  object ErrorResponse {
+
+    case object NotFound extends ErrorResponse(None)
+
+    final case class Message(message: String) extends ErrorResponse(message.some)
+
+  }
```
Comment (stsatlantis) [inline src/main/scala/com/wanari/renervator/api/WakerApi.scala:79]: Helps to return proper type for the API

File: src/test/scala/com/wanari/renervator/api/WakerApiSpec.scala [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        val wakerApi = new WakerApi(
+          database(List(Host("localhost", "localhost", "", isOnline = false, 1L))),
+          new MockedNetworkingService(Left("I'm failed"))
+        )
+
+        val request = POST(IdDTO(1L).asJson, Uri.uri("/wakeItUp")).unsafeRunSync
+        val response: Option[Response[IO]] = wakerApi.route.run(request).value.unsafeRunSync()
+        response match {
+          case Some(response) =>
+            response.status shouldBe Status.InternalServerError
```
Comment (stsatlantis) [inline src/test/scala/com/wanari/renervator/api/WakerApiSpec.scala:56]: add assertation on body

File: src/main/scala/com/wanari/renervator/Database.scala [hunk 3]
```diff
@@ -13,7 +13,7 @@ class Database {
 
   def get(id: Long): IO[Option[Host]] = {
     IO{
-      Try(map.get(id)).toOption
+      Try(Option(map.get(id))).toOption.flatten
```
Comment (tg44) [inline src/main/scala/com/wanari/renervator/Database.scala:16]: why? 
The `Try[Option[?]].toOption.flatten` is not the same as `Try[?].toOption`?

File: src/main/scala/com/wanari/renervator/api/WakerApi.scala [hunk 4]
```diff
@@ -1,40 +1,71 @@
...
+    )
+
+  val hostListApi: Endpoint[Unit, Unit, HostListDTO, Nothing] = endpoint
+    .get
+    .in("wol")
+    .out(jsonBody[HostListDTO])
+
+  import tapir.server.http4s._
+
+  val wakeItUpApiLogic: IdDTO => IO[Either[ErrorResponse, Json]] = (idWrapper: IdDTO) => (for {
```
Comment (tg44) [inline src/main/scala/com/wanari/renervator/api/WakerApi.scala]: sooo this is a def?

File: src/main/scala/com/wanari/renervator/api/WakerApi.scala [hunk 5]
```diff
@@ -1,40 +1,71 @@
...
+    host <- EitherT.fromOptionF[IO, ErrorResponse, Host](database.get(idWrapper.id), ErrorResponse.NotFound)
+      _ <- EitherT(networkingService.sendMagicPocket(host)).leftMap[ErrorResponse](ErrorResponse.Message)
+      result = Json.obj()
+  } yield result).value
+
+  val hostListLogic: Unit=> IO[Either[Unit, HostListDTO]] = (_: Unit) => database.all.map[Either[Unit, HostListDTO]](e => Right(HostListDTO(e)))
```
Comment (tg44) [inline src/main/scala/com/wanari/renervator/api/WakerApi.scala]: and this is a def too

File: src/test/scala/com/wanari/renervator/api/WakerApiSpec.scala [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+            response.as[Json].unsafeRunSync() shouldBe HostListDTO(List(expectedHosts)).asJson
+          case _ => fail()
+        }
+      }
+
+    }
+  }
+}
+
+class MockedNetworkingService(result: Either[String, Unit]) extends NetworkingService {
```
Comment (tg44) [inline src/test/scala/com/wanari/renervator/api/WakerApiSpec.scala]: this can go to a separate file under the testutils or fakes or mocks or sth like that

**Review by tg44 — COMMENTED** [review]
Thx the contrib and also the tests!

Are there any reasons why you prefer `val sth: A =>B` over `def sth(a: A): B` ?

Also, can you open up a swagger endpoint too?

File: src/main/scala/com/wanari/renervator/Database.scala (hunk 3 above)
Comment (tg44) [reply src/main/scala/com/wanari/renervator/Database.scala:16]: ohhhh I get it now... It's a java map, and the get is not the same as on scala maps. My bad!

**Review by stsatlantis — COMMENTED** [review]
> Are there any reasons why you prefer `val sth: A =>B` over `def sth(a: A): B` ?

`compose` and `andThen` are specified for Function and not for method. For me the `serverLogic` expects a function :)
 
> Also, can you open up a swagger endpoint too?
created an issue for that: https://github.com/TeamWanari/renervator-wol/issues/14

**Comment by stsatlantis** [comment]
I would wait for https://github.com/TeamWanari/renervator-wol/pull/13 to be merged to include it as well.

**Review by stsatlantis — COMMENTED** [review]
From my side,  I consider the PR done. If you need anything let me know.

File: src/test/scala/com/wanari/renervator/api/WakerApiSpec.scala (hunk 6 above)
Comment (stsatlantis) [reply src/test/scala/com/wanari/renervator/api/WakerApiSpec.scala]: moved under _root_.util package

**Comment by tg44** [comment]
Thanks the contrib!

**Pull request merged**